provider/infoblox.go

```diff
@@ -0,0 +1,389 @@
+			}
+			return
+		}
+		// Ensure the record type is suitable
+		changeMap[zone.Fqdn] = append(changeMap[zone.Fqdn], change)
+	}
+
+	for _, change := range changes.Create {
+		mapChange(created, change)
+	}
+	for _, change := range changes.Delete {
+		mapChange(deleted, change)
+	}
+	for _, change := range changes.UpdateNew {
+		mapChange(updated, change)
+	}
+	for _, change := range changes.UpdateOld {
+		mapChange(deleted, change)
+	}
+
+	return created, deleted, updated
+}
+
+func (p *InfobloxProvider) findZone(zones []ibclient.ZoneAuth, name string) *ibclient.ZoneAuth {
+	var result *ibclient.ZoneAuth
+
+	// Go through every zone looking for the longest name (i.e. most specific) as a matching suffix
+	for idx := range zones {
+		zone := &zones[idx]
+		if strings.HasSuffix(name, zone.Fqdn) {
+			if result == nil || len(zone.Fqdn) > len(result.Fqdn) {
+				result = zone
+			}
+		}
+	}
+	return result
+}
+
+func (p *InfobloxProvider) newObject(e *endpoint.Endpoint, getRef bool) (obj ibclient.IBObject, ref string, err error) {
+	switch e.RecordType {
+	case endpoint.RecordTypeA:
+		var records []ibclient.RecordA
+		obj = ibclient.NewRecordA(
+			ibclient.RecordA{
+				Name:     e.DNSName,
+				Ipv4Addr: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	case endpoint.RecordTypeCNAME:
+		var records []ibclient.RecordCNAME
+		obj = ibclient.NewRecordCNAME(
+			ibclient.RecordCNAME{
+				Name:      e.DNSName,
+				Canonical: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	case endpoint.RecordTypeTXT:
+		var records []ibclient.RecordTXT
+		if target, err := strconv.Unquote(e.Target); err == nil && !strings.Contains(e.Target, " ") {
+			e.Target = target
+		}
+		obj = ibclient.NewRecordTXT(
+			ibclient.RecordTXT{
+				Name: e.DNSName,
+				Text: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	default:
+		return nil, "", fmt.Errorf("unsupported record type '%s'", e.RecordType)
+	}
+}
+
+func (p *InfobloxProvider) createRecords(created infobloxChangeMap) {
+	for zone, endpoints := range created {
+		for _, endpoint := range endpoints {
+			if p.dryRun {
+				logrus.Infof(
+					"Would create %s record named '%s' to '%s' for Infoblox DNS zone '%s'.",
+					endpoint.RecordType,
+					endpoint.DNSName,
+					endpoint.Target,
+					zone,
+				)
+				continue
+			}
+
+			logrus.Infof(
+				"Creating %s record named '%s' to '%s' for Infoblox DNS zone '%s'.",
+				endpoint.RecordType,
+				endpoint.DNSName,
+				endpoint.Target,
+				zone,
+			)
+
+			obj, _, err := p.newObject(endpoint, false)
+			if err == nil {
```

Comment:
i think more natural would be to "fail early" in this case, i.e. 

```go
if err != nil { 
      // return error 
}
_, err = p.client.CreateObject(obj)
```

Reply:
same for  `update` and `delete` methods

Reply:
You are so right. I have forsaken one of the core principles.